File: C++/bubblesort.cpp
#include <iostream>
#include <vector>
#include <stdlib.h>

template <typename T>
void swap(T& a, T& b) {
    T tmp = b;
    b = a;
    a = tmp;    
}

template <typename T>
void bubblesort(T& vec) {
    auto f = vec.begin();
    auto s = vec.begin();
    ++s;
    bool swapped = false;
    while (s != vec.end()) {
	if (*f > *s) {
	    swap(*f, *s);
	    swapped = true;
	}
	++f; ++s;
    }
    if (swapped) {
	bubblesort(vec);
    }
}

int main() {

    std::vector<int> v;

    for (int x = 0; x < 1000000; ++x) {
	v.push_back(rand() % 1000000);
    }
    
    bubblesort(v);
    return 0;
}

File: Python/bubblesort.py
import random
import sys

sys.setrecursionlimit(20000000)


def bubblesort(vec):
    idx = 2
    swapped = False
    while (idx < len(vec)):
        if (vec[idx-1] > vec[idx]):
            vec[idx-1], vec[idx] = vec[idx], vec[idx-1]
            swapped = True
        idx += 1
    if swapped:
        bubblesort(vec)

if __name__ == '__main__':
    vec = []
    for x in range(1000000):
        vec.append(random.randint(0,x))
    
    bubblesort(vec)
    print vec

    
